src/langmem/short_term/summarization.py

```diff
@@ -373,6 +373,301 @@ def call_model(state):
+            messages=(
+                messages
+                if existing_system_message is None
+                else [existing_system_message] + messages
+            ),
+        )
+
+    # Get previously summarized messages, if any
+    summarized_message_ids = set()
+    total_summarized_messages = 0
+    existing_summary = running_summary
+    if running_summary:
+        summarized_message_ids = running_summary.summarized_message_ids
+        # Adjust the summarization token budget to account for the previous summary
+        max_tokens_to_summarize -= token_counter(
+            [SystemMessage(content=running_summary.summary)]
+        )
+        # If we have an existing running summary, find how many messages have been
+        # summarized so far based on the last summarized message ID.
+        for i, message in enumerate(messages):
+            if message.id == running_summary.last_summarized_message_id:
+                total_summarized_messages = i + 1
+                break
+
+    # We will use this to ensure that the total number of resulting tokens
+    # will fit into max_tokens window.
+    total_n_tokens = token_counter(messages[total_summarized_messages:])
+
+    # Go through messages to count tokens and find cutoff point
+    n_tokens = 0
+    idx = max(0, total_summarized_messages - 1)
+    # map tool call IDs to their corresponding tool messages
+    tool_call_id_to_tool_message: dict[str, ToolMessage] = {}
+    should_summarize = False
+    n_tokens_to_summarize = 0
+    for i in range(total_summarized_messages, len(messages)):
+        message = messages[i]
+        if message.id is None:
+            raise ValueError("Messages are required to have ID field.")
+
+        if message.id in summarized_message_ids:
+            raise ValueError(
+                f"Message with ID {message.id} has already been summarized."
+            )
+
+        # Store tool messages by their tool_call_id for later reference
+        if isinstance(message, ToolMessage) and message.tool_call_id:
+            tool_call_id_to_tool_message[message.tool_call_id] = message
+
+        n_tokens += token_counter([message])
+
+        # Check if we've reached max_tokens_to_summarize
+        # and the remaining messages fit within the max_remaining_tokens budget
+        if (
+            n_tokens >= max_tokens_before_summary
+            and total_n_tokens - n_tokens <= max_remaining_tokens
+            and not should_summarize
+        ):
+            n_tokens_to_summarize = n_tokens
+            should_summarize = True
+            idx = i
+
+    # Note: we don't return here since we might still need to include the existing summary
+    if not should_summarize:
+        messages_to_summarize = []
+    else:
+        messages_to_summarize = messages[total_summarized_messages : idx + 1]
+
+    # If the last message is an AI message with tool calls,
+    # include subsequent corresponding tool messages in the summary as well,
```

Comment:
any way we could share all this code? Seems mostly the same sans 1 or 2 `ainvoke`'s?

Reply:
yep - was mostly worried about readability but it is a _lot_ of duplication, so i think worth factoring out